### PR TITLE
Name resolver

### LINKED ISSUE #51: Switch curie lookup over to name resolver
http://robokop.renci.org:2433/docs#/lookup/lookup_curies_lookup_post

### PR DESCRIPTION
**Pull request by alongreyber**
Switched node search to use nameResolver and nodeNormalizer for lookup instead of ranker. Closes #51.

### CODE REVIEW

File: frontend/src/pages/question/panels/NodePanel.jsx [hunk 1]
```diff
@@ -151,22 +150,77 @@ export default function NodePanel({ panelStore }) {
...
+      if (r.status === 'error') {
+        nodeNormalizerError = r;
+      }
+    });
+
+    if (nodeNormalizerError === 'error') {
+      displayAlert('error',
+        'Failed to contact node normalizer to search curies. You will still be able to select generic types. Please try again later');
+      node.setLoading(false);
+      return;
```
Comment (maximusunc) [inline frontend/src/pages/question/panels/NodePanel.jsx:211]: This seems like it won't work. We're setting nodeNormalizerError to the whole response object if its status is 'error', then checking if that object is equal to the string 'error'. It won't. The forEach can also be a find() so we stop as soon as we get an error. Cause on line 211, we stop and return if there are any errors. Are there any chances one call to the normalizer will fail while others work? Do we want to handle that case, or just show an error if any of them fail?

File: frontend/src/pages/question/panels/NodePanel.jsx [hunk 2]
```diff
@@ -151,22 +150,77 @@ export default function NodePanel({ panelStore }) {
...
+      if (newLength < maxUrlLength) {
+        existingArray.push(c);
+      } else {
+        curiesChunked.push([c]);
+      }
+    });
+
+    const normalizationAPICallResponses = await Promise.all(
+      curiesChunked.map((cs) => API.nodeNormalization.getNormalizedNodes(cs)),
+    );
```
Comment (maximusunc) [inline frontend/src/pages/question/panels/NodePanel.jsx:197]: This all looks reasonable. But silly. We should open an issue or something on the Normalizer repo and see if they can accept a POST with body so we don't have to do this.

File: frontend/src/pages/question/panels/NodePanel.jsx [hunk 3]
```diff
@@ -151,22 +150,77 @@ export default function NodePanel({ panelStore }) {
...
+        'Failed to contact node normalizer to search curies. You will still be able to select generic types. Please try again later');
+      node.setLoading(false);
+      return;
     }
+
+    const curiesWithInfo = {};
+    normalizationAPICallResponses.forEach((r) => Object.assign(curiesWithInfo, r));
+
+    node.updateCuries(
+      Object.values(curiesWithInfo).filter((c) => c).map((c) => ({
```
Comment (maximusunc) [inline frontend/src/pages/question/panels/NodePanel.jsx:218]: What is this filter for?

**Review by maximusunc — CHANGES_REQUESTED** [review]
Our routing solutions seem very haphazard. Is there a disadvantage to just passing everything through the proxy so we don't have to worry about CORS?

File: frontend/src/pages/question/panels/NodePanel.jsx (hunk 2 above)
Comment (patrickkwang) [reply frontend/src/pages/question/panels/NodePanel.jsx:197]: The fix is already in the works, but let's go ahead with a workaround in case they're slow.

**Review by maximusunc — CHANGES_REQUESTED** [review]
It looks like you have a missing trailing comma in nameResolver from the eslint check. I would also like a reply to my overall comment from my first review.

**Comment by alongreyber** [comment]
> Our routing solutions seem very haphazard. Is there a disadvantage to just passing everything through the proxy so we don't have to worry about CORS?

The disadvantages I can think of:
- Possibly worse performance for the client because there may be a bottleneck with the proxy if we send out a lot of requests at once
- Higher load on the server

These disadvantages are definitely small. Switching to redirecting everything through the proxy would save us some complexity, but not a significant amount because of the way the code is written. I think the advantages are small as well. Overall, definitely a toss-up, but I would lean towards not redirecting everything through the proxy just because we already have it written this way.

**Comment by patrickkwang** [comment]
If it's more or less a neutral change and things are working right now, I recommend leaving it alone until/unless it feels more urgent in the future.

File: proxy/nginx.conf [hunk 4]
```diff
@@ -24,6 +24,10 @@ http {
 		location /api/robokache/ {
 			proxy_pass http://robokache/api/;
 		}
+
+		location /api/external/nodeNormalization/ {
+			proxy_pass https://nodenormalization-sri.renci.org/;
```
Comment (maximusunc) [inline proxy/nginx.conf:29]: This should probably be a variable, like we did with the frontend url, so it doesn't crash if this host isn't up yet.

**Review by maximusunc — CHANGES_REQUESTED** [review]
I'm not sure if this is just my janky docker situation, but I'm getting 404s from the normalizer for everything. Or maybe the service is down right now.

I don't know if we want to do anything, but when the proxy service logs these routes out in the running console, it includes the whole route, which is massive since we're including every curie in the actual route, so the log is super swamped.

**Review by maximusunc — APPROVED** [review]

**Pull request merged**